core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java

```diff
@@ -58,26 +61,38 @@ public DataType call() throws Exception {
         HystrixRequestContext existingState = HystrixRequestContext.getContextForCurrentThread();
         boolean success = false;
         try {
-            RequestContext.initialize(parentContext);
-            ServiceContext.initialize(parentServiceContext);
-            HystrixRequestContext.setContextOnCurrentThread(parentThreadState);
-            // Parent thread span info is passed onto Datasource thread using Brave's ThreadLocal implementation
-            if (serverSpan != null && serverSpan.getSpan() != null) {
-                Brave.getServerSpanThreadBinder().setCurrentSpan(serverSpan);
-            }
-            startTrace(dataSource, request);
+            initAllContext();
             DataType dataType = dataSource.call();
             success = true;
             return dataType;
         } finally {
-            endTrace(dataSource, success);
-            RequestContext.shutDown();
-            ServiceContext.shutDown();
-            HystrixRequestContext.setContextOnCurrentThread(existingState);
-            Brave.getServerSpanThreadBinder().setCurrentSpan(null);
+            shutdownAllContext(existingState, success);
         }
     }
 
+    private void initAllContext() {
```

Comment:
No we don't have a base class for filter & datasource. Remember our earlier discussion of defining Block in lego. Till that's done, this is not doable.